File: source/TrafficLight.c
#include <stdio.h>
#include <wiringPi.h>
#include <signal.h>
#include <string.h>


//BCM GPIO connection
const int RED    = 16;
const int YELLOW = 20;
const int GREEN  = 21;
static sig_atomic_t end = 0;

static void sighandler(int signo) {
  end = 1;
}

void SetRYG(int Red, int Yellow, int Green) {
	digitalWrite(RED,    Red   );
	digitalWrite(YELLOW, Yellow);
	digitalWrite(GREEN,  Green );	
}

void WaitMS(int MS) {
	if (!end)	delay(MS);
}

int main(void) {
	struct sigaction sa;

	setbuf(stdout, NULL); // deactivate stdout buffering 
	if (wiringPiSetupGpio() == -1) { // Init with BCM GPIO numbers
		perror("wiringPiSetup failed");
		return 1;
	}
	memset(&sa, 0, sizeof(struct sigaction)); //Handling stop events, like ctrl+c 
	sa.sa_handler = sighandler;
	sigaction(SIGINT, &sa, NULL);
	sigaction(SIGQUIT,&sa, NULL);
	sigaction(SIGTERM,&sa, NULL);

	printf("Raspberry Pi traffic light (press ctrl+c to quit)\n");
	pinMode(RED,    OUTPUT);
	pinMode(YELLOW, OUTPUT);
	pinMode(GREEN,  OUTPUT);
	SetRYG(LOW, LOW, LOW);
	while(!end) {
		SetRYG(HIGH, LOW, LOW);
		WaitMS(5000);
		SetRYG(HIGH, HIGH, LOW);
		WaitMS(1000);
		SetRYG(LOW, LOW, HIGH);
		WaitMS(4000);
		for (int nBlinkCount=0; nBlinkCount<3; nBlinkCount++) { 
			if (end) break;
			SetRYG(LOW, LOW, LOW);
			WaitMS(1000);
			SetRYG(LOW, LOW, HIGH);
			WaitMS(1000);
		}
		SetRYG(LOW, HIGH, LOW);
		WaitMS(2000);
	}
	SetRYG(LOW, LOW, LOW);
	pinMode(RED,    INPUT);
	pinMode(YELLOW, INPUT);
	pinMode(GREEN,  INPUT);
	return 0;
}

File: source/TrafficLight.cs
namespace Raspjamming
{
	using System;
	using System.Threading;

	static class TrafficLight
	{
		const int Red = 16;
		const int Yellow = 20;
		const int Green = 21;
		private static volatile bool _end = false;

		static void Console_CancelKeyPress(object sender, ConsoleCancelEventArgs e)
		{
			e.Cancel = true;
			_end = true;
		}

		static void SetRYG(int red, int yellow, int green)
		{
			WiringPi.DigitalWrite(Red, red);
			WiringPi.DigitalWrite(Yellow, yellow);
			WiringPi.DigitalWrite(Green, green);
		}

		static void WaitMS(int ms)
		{
			if (!_end)
			{
				Thread.Sleep(ms);
			}
		}

		static int Main()
		{
			Console.CancelKeyPress += Console_CancelKeyPress;
			// Init with BCM GPIO numbers
			if (WiringPi.WiringPiSetupGpio() == -1)
			{
				Console.WriteLine("wiringPiSetup failed");
				return 1;
			}
			Console.WriteLine("Raspberry Pi traffic light (press ctrl+c to quit)");
			WiringPi.PinMode(Red, WiringPi.Output);
			WiringPi.PinMode(Yellow, WiringPi.Output);
			WiringPi.PinMode(Green, WiringPi.Output);
			SetRYG(WiringPi.Low, WiringPi.Low, WiringPi.Low);
			while (!_end)
			{
				SetRYG(WiringPi.High, WiringPi.Low, WiringPi.Low);
				WaitMS(5000);
				SetRYG(WiringPi.High, WiringPi.High, WiringPi.Low);
				WaitMS(1000);
				SetRYG(WiringPi.Low, WiringPi.Low, WiringPi.High);
				WaitMS(4000);
				for (int blinkCount = 0; blinkCount < 3; blinkCount++)
				{
					if (_end)
					{
						break;
					}
					SetRYG(WiringPi.Low, WiringPi.Low, WiringPi.Low);
					WaitMS(1000);
					SetRYG(WiringPi.Low, WiringPi.Low, WiringPi.High);
					WaitMS(1000);
				}
				SetRYG(WiringPi.Low, WiringPi.High, WiringPi.Low);
				WaitMS(2000);
			}
			SetRYG(WiringPi.Low, WiringPi.Low, WiringPi.Low);
			WiringPi.PinMode(Red, WiringPi.Input);
			WiringPi.PinMode(Yellow, WiringPi.Input);
			WiringPi.PinMode(Green, WiringPi.Input);
			return 0;
		}
	}
}
